### DISCUSSION TITLE
Hajk 3.8 Release Candidate 1

### DISCUSSION THREAD

**Discussion by jacobwod** [hajkmap/Hajk#907] [General]
_[212 commits](https://github.com/hajkmap/Hajk/compare/v3.7.0...v3.8.0-rc.1) and changes in 130 files since the 3.7 release – here's Hajk 3.8! This release was brought to you by @Hallbergs  and @jacobwod . Please see the list of commits for all contributors._

# Changelog

### Added FME Server plugin. #750
This opens up new possibilities for integration with Safe's FME Server application. 
![MicrosoftTeams-image](https://user-images.githubusercontent.com/110222/141750991-01aca969-cc87-4567-8983-070dcecee5b7.png)

### Search plugin respects different response types. #882
Search plugin now respects different info formats (GML 2/3/3.2 and GeoJSON). Makes Hajk work with QGIS Server's WFS implementation. 

### Improvements in Edit. #884
More data types are allowed in the plugin.

### Improve Hajk's accessibility. #877
It's now possible to navigate the map using keyboard. 

### Make Client UI user-aware. #899
The added UI Control is just a proof-of-concept, with virtually no functionality other than displaying the user's name in the UI. However, this opens up a lot of possibilities for plugins to optionally read provided information from Active Directory. 

## New features

* Major improvements for **QGIS Server** users, including better support for `GetLegendGraphics` (#879) and `FeatureInfo` (#880) requests. 
* **TLS** Support in **Backend** makes it possible to authenticate via **LDAPS**. #893
* Allow Admin to enable Home button #876
* **Optional info** is now shown also for **background layers**. #808
* Make it possible to disable View rotation, among other **default OpenLayers interactions** #875
* **Location plugin**: make it more like other map apps #863
* Better **EDP integration** support #861
* Added endpoint **`/fmeproxy`** as a new controller https://github.com/hajkmap/Hajk/pull/900

## Bug fixes and minor features
*  #869, #864, #854, #860, #859, #894, #898, #897, #901, #902.

# Checklist for admins
- [ ] Verify that search works with different kinds of sources. **Previous to this release, search always used GeoJSON, but as of this release the setting from Admin UI is respected. Since the default setting in Admin UI has always been GML2, _that's what will be used by default from this release_. If you want to stay with GeoJSON: make sure to change the response type for your search sources.**
- [ ] Verify that FeatureInfo ("info click") works, for multiple layer types.
- [ ] Ensure that the Edit plugin works with your current configuration. The plugin has seen substantial update in this release. 
- [ ] If you use the Active Directory authentication in NodeJS backend, you now have the possibility to supply certificates and use LDAPS. Check out details in `.env`. 
- [ ] If you're planning to use the new FME Server plugin, make sure that the `url_base` key is set in Admin UI's `config.json`. (This has been added to `config.json` in this release.)
- [ ] If you'd like to use the more compact version of the Location plugin, ensure to set the `target` properly to `control` in Admin UI for that plugin. 

---

# Downloads
There are multiple ways to get the latest Hajk up and running, including Docker container.

## Pre-compiled releases
This time we have **three different** version of the pre-compiled solution (scroll down for download links):
1. [Hajk 3.8.0-rc.1 NodeJS edition](https://github.com/hajkmap/Hajk/releases/download/v3.8.0-rc.1/hajk-v3.8.0-rc.1-nodejs.zip)
1. [Hajk 3.8.0-rc.1 Simple edition](https://github.com/hajkmap/Hajk/releases/download/v3.8.0-rc.1/hajk-v3.8.0-rc.1-simple.zip)
1. [Hajk 3.8.0-rc.1 .NET edition](https://github.com/hajkmap/Hajk/releases/download/v3.8.0-rc.1/hajk-v3.8.0-rc.1-dotnet.zip)

## Docker container
The container is available on Docker Hub: https://hub.docker.com/r/hallbergs/hajk

# Installation
Refer to our recently updated [installation guide](https://github.com/hajkmap/Hajk/wiki/Installation-guide-%28for-pre-packaged-releases%29).

<hr /><em>This discussion was created from the release <a href='https://github.com/hajkmap/Hajk/releases/tag/v3.8.0-rc.1'>Hajk 3.8 Release Candidate 1</a>.</em>

**Comment by perosoderstrom** [comment]
Problem with "Utför sökning" in search plugin.

This works as expected:

Install Hajk 3.8 Release Candidate 1, nodejs backend, and configure .env, client/appConfig.json and admin/config.json. Hajk starts OK.

If you do a search, like "_storgatan_", you get a selection of addresses in Gothenburg, as expected, and if you click on "**Utför sökning**"  you get the rest, as expected.

But if I clone and install from source I get problem:

Git clone, git checkout develop, no changes in config files, npm install, npm run dev/npm start: Hajk starts OK.

If you do a search, like "_storgatan_" you get a selection of addresses in Gothenburg, as expected, and if you click on "Utför sökning"  you get all, as expected.

If I change "mapserviceBase": "http://localhost:3002/api/v1", in appConfig.json and try the same, then **"Utför sökning**" says **_"Sökningen gav inget resultat"_**.

**Reply by Hallbergs** [reply]
Hmm. This sounds really strange. I've seen that the wfs used in the standard configurations (Addresses in Gothenburg) have been returning a connection error _sometimes_. Could it be that the wfs just happened to return a error? Could you check the response in the network tab when performing a search so we can track down the culprit?

**Reply by jacobwod** [reply]
This one turned out to be a local issue where @perosoderstrom had `matchCase` active, right? I assume it's "resolved".

**Reply by perosoderstrom** [reply]
Yes, this is resolved.

**Comment by perosoderstrom** [comment]
I get OK response (Status 200) from the fetch, but no features:

![Skärmbild 2021-11-16 080429](https://user-images.githubusercontent.com/51197693/141940311-b9ae89b3-d3e3-47f3-ae97-fed018b01c2c.png)
![Skärmbild 2021-11-16 080405](https://user-images.githubusercontent.com/51197693/141940381-b3e813a1-c06b-40d3-bf70-512dd53abf36.png)

More odd things:
1. In admin -"Söktjänster", if I click on the original layer "Adresser Göteborg", I get an error. After the click i can still see the form, and I see that the layer name differs in "Lagerlista" and "Valt lager", and the layer in "Lagerlista" is not selected.
![Skärmbild 2021-11-16 082129](https://user-images.githubusercontent.com/51197693/141940760-c5f1580f-be1d-42be-9701-eb18b7bdccd1.png)
![Skärmbild 2021-11-16 082357](https://user-images.githubusercontent.com/51197693/141940792-bbeea2fd-81fb-4bbb-8de7-6b7eebff90b0.png)

2. If I delete "Adresser Göteborg" from search services in admin, the layer is still left in map_1.json.

**Reply by jacobwod** [reply]
What does the the fetch request look like in detail? You should be able to "Copy as cURL" in your browser's developer tools. Here's what [mine](https://hajk.jwod.se) when Hajk does an autocomplete lookup for the term `storg`:
```
curl 'https://opengeodata.goteborg.se/services/adresser/wfs/v1' \
-X 'POST' \
-H 'Origin: https://hajk.jwod.se' \
-H 'Accept: */*' \
-H 'Content-Type: text/xml' \
-H 'Pragma: no-cache' \
-H 'Cache-Control: no-cache' \
-H 'Accept-Language: sv-se' \
-H 'Host: opengeodata.goteborg.se' \
-H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.1 Safari/605.1.15' \
-H 'Content-Length: 552' \
-H 'Accept-Encoding: gzip, deflate, br' \
-H 'Connection: keep-alive' \
--data-binary '<GetFeature xmlns="http://www.opengis.net/wfs" service="WFS" version="1.1.0" outputFormat="GML2" maxFeatures="100" xsi:schemaLocation="http://www.opengis.net/wfs http://schemas.opengis.net/wfs/1.1.0/wfs.xsd" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"><Query typeName="adresser.wfs.v1:adresser" srsName="EPSG:3006"><Filter xmlns="http://www.opengis.net/ogc"><PropertyIsLike wildCard="*" singleChar="." escapeChar="!" matchCase="false"><PropertyName>adress</PropertyName><Literal>storg*</Literal></PropertyIsLike></Filter></Query></GetFeature>'
```

I'm running directly from the Docker build, there's an [Admin UI](https://hajk.jwod.se/admin/) over here if you want to compare the search service's setup.

**Reply by Hallbergs** [reply]
I just had a meeting with Per. The issue was that `matchCase` was saved as `true` in localStorage, leading to no hits :) 

The admin part we must investigate!

**Comment by jacobwod** [comment]
We have multiple problems with the Edit plugin. It does not work with our previous configurations, and we haven't yet managed to configure it "correctly" in the new Admin UI either. 

**Has anyone had any success with saving from WFS in 3.8 yet?** @perosoderstrom @hitomin21 @Hallbergs @eri-blan @WCarlsson

_Edit: I've managed to save at least one of our GeoServer-served WFS services: a PostGIS table with Point geometries. Still extensive testing is needed though, but it's good to see that some existing configuration seems to work!_

**Reply by jesade-vbg** [reply]
This bug fix resulted in a bug that causes the plugin to crash (not related to save):
https://github.com/hajkmap/Hajk/commit/4b4b47891a7777386eaaf7213bf4f29139778514#r60362119

**Reply by jacobwod** [reply]
https://github.com/hajkmap/Hajk/commit/4b4b47891a7777386eaaf7213bf4f29139778514#commitcomment-60363003

**Reply by jacobwod** [reply]
It's resolved and fix pushed into `master`.

**Reply by Hallbergs** [reply]
I've tested the edit plugin in out environment and it seems to work!

**Reply by hitomin21** [reply]
I tested in our environment now and it seems to work based on what I can see. 

Does anyone have an error message or some description for cases where old configurations did not work?

Also, for those that have been unable to make save the edit tool work, do you have any details about that? I and @jesade-vbg have tested and it works for us

**Reply by WCarlsson** [reply]
Edit doesn´t seem entirely reliable for us yet, we will try to test more in depth in order to find something specific to point to.

**Reply by WCarlsson** [reply]
@hitomin21 not problems with old configurations but these are the problems that we are still experiencing with edit:

While editing an existing object any attribute values of null is filled with [object Object] in client, it is possible to simply remove this text and fill with desirable value or leave empty to achieve a succesful commit.

Date attributes does not work, it is probably a problem with oracle and a difference in format. A date attribut from oracle can be seen in network tab with format "2021-10-25 00:00:00" but does not show in edit in client. An WFS-T call shows that hajk has date format "2021-11-12T16:06" and results in error "SQL: ORA-01830: date format picture ends before converting entire input string"

Non-editable attributes does not show values and simply states the attribute name, not sure if this is intended.

**Reply by jesade-vbg** [reply]
I have no issues saving from WFS.

**Reply by jacobwod** [reply]
Here's a comprehensive description, @jesade-vbg, @hitomin21 : #927.

**Reply by hitomin21** [reply]
@WCarlsson Non-editable values should only show a value if you specified a default value in admin, so the behavior you saw might be expected depending on the configuration.

I will see what I can do about that Oracle error, if we can get values into postgis with the same format as Oracle. If that does not work, we might have to add some database type to admin. It would be better if we could somehow modify the format using QGIS server or Geoserver if it comes to that.

**Reply by hitomin21** [reply]
We are testing fixes for object Object in another discussion
https://github.com/hajkmap/Hajk/issues/927

**Reply by hitomin21** [reply]
@WCarlsson It seems like changing the format to 2021-10-25 00:00:00 will make postgis discard the time part of the datetime so we either need a setting for whether Oracle is used (which I prefer not to implement to keep the code as database agnostic as possible) or allow to store datetimes in string columns instead.

I think putting them in string columns ( is better, that should work across all database servers). What do you think? It means that you would specify "Datum & Tid" under Typ for a string field (same as we allow dates for that).
![bild](https://user-images.githubusercontent.com/22234673/144249611-4c78a7c8-05a8-44cc-ae00-56bc8fbdcb7c.png)

**Reply by hitomin21** [reply]
If that sounds OK, I will add it

**Reply by jacobwod** [reply]
I can't find any reference (if anyone else can point out the correct authoritative reference, feel free to paste the link!) at the moment. But if I recall correctly, the WFS-T will use the `xsd:date` format as defined in the XML specification. 

I imagine that if the WFS-T requests sets the date value as a correct `xsd:date` (rather than `xsd:string`), the WFS server should take care of the rest. We are only interested in sending the data correctly to the WFS service (following the standard's specification) – whatever happens next (whether it's writing to PostGIS, Oracle or anything else) is the responsibility of the WFS server. 

Can you check the POST to WFS-T @WCarlsson and verify that the date being sent is in `xsd:date`? 

@hitomin21 I'd rather see that we don't recommend storing datetimes in string fields. Let's try to follow the standard: a date is a date, not a string. Let's await @WCarlsson's response first. 👍

**Reply by hitomin21** [reply]
@WCarlsson : Just a friendly reminder to get back about the WFS-T response.

**Reply by WCarlsson** [reply]
@hitomin21 Thanks for the reminder, this slipped my mind!
It seems that QGIS always treats oracle date format as dateTime per default and with dateTime there seems to be some problem, I will try to investigate this further to see if I can get dateTime to work or more info on why it doesn´t.

If we make sure to specify the attribute as a date attribute when setting up the service in QGIS instead of letting it assume the datatype by itself it is recognised as a date and then the edit plugin works as expected. Thanks for the tip to verify this @jacobwod

**Comment by Hallbergs** [comment]
Just pushed a fix to master which fixes a bug in the XLSX-export. (We were trying to get the keys from the features with .properties, but on ol-features we must use .getProperties()).

**Reply by Hallbergs** [reply]
(The xlsx export in the search module)

**Reply by Hallbergs** [reply]
Just pushed another fix for the xlsx-export. Now the geometry-field is removed from the export, and the values should be aligned even if some properties are missing from some features.

**Reply by perosoderstrom** [reply]
Very good @Hallbergs! This solves en old issue ( #411)

**Reply by Hallbergs** [reply]
Wonderful!

**Reply by jacobwod** [reply]
Great find and fix @Hallbergs

**Reply by Hallbergs** [reply]
We might still miss keys... I'll investigate further. @jacobwod

**Reply by jacobwod** [reply]
I've added an issue as I think it's outside of scope for this thread for now. #916.

**Comment by perosoderstrom** [comment]
Problem saving layer search properties in admin with .Net Backend

If you configure search properties, e.g. search URL, search field etc., for a WMS layer the properties are not saved in layers.json. 

![Skärmbild 2021-11-17 141621](https://user-images.githubusercontent.com/51197693/142207938-482de96b-e852-4f2c-a21d-2b605336f0ff.png)


This works OK in Node Backend.

![Skärmbild 2021-11-17 141652](https://user-images.githubusercontent.com/51197693/142207982-4f6d70d3-91b6-4a5b-b40f-9eb0bf76aa01.png)

**Reply by Hallbergs** [reply]
Nice catch! Will definitely look into that!

**Reply by Hallbergs** [reply]
0668b4bffe17cd5e5511fd20012c1efc417d6d50 should have sorted this. (Just pushed to master)

**Reply by perosoderstrom** [reply]
Just tested this and it seems to work ok. Thanks, @Hallbergs!

**Reply by perosoderstrom** [reply]
Noticed another problem (not having to do with this)  when testing this.
In` /mapservice/App_Data/map_1.json`, map zoom is set to 1.33 (**"zoom": 1.33,**), and dotnet backend requires an integer. Setting this to "**zoom: 2**" will solve it.

**Reply by Hallbergs** [reply]
It's taken care of now.

**Comment by WCarlsson** [comment]
We have found a couple of problems.
Baselayers (bakgrundslager) made from multiple layers from a wms-service doesn´t work anymore, react errors show in console.

In edit a zoom occurs when closing with a "fast" doubleclick, with a slower doubleclick this effect is avoided. Seems similar to https://github.com/hajkmap/Hajk/issues/902

When adding a search or edit service from a QGIS server we have to add ?SERVICE=WFS to the URL in order to be able to show attributes in admin. If we add the service parameter the service seems to work but calls from the edit service adds a second service parameter. In order to have the same format of URL for layer, search and edit it would be nice to not have to add the service parameter.

**Reply by Hallbergs** [reply]
I just pushed a fix for the zoom-issue (e96c45918a3eaf39b1792678c81744f6a8e2331a).

**Reply by jacobwod** [reply]
> Baselayers (bakgrundslager) made from multiple layers from a wms-service doesn´t work anymore, react errors show in console.

This _could_ be related to the recent addition of "Show more info" to `BackgroundSwitcher`. @hitomin21 or @jesade-vbg, can you take a look?

> When adding a search or edit service from a QGIS server we have to add ?SERVICE=WFS to the URL in order to be able to show attributes in admin. If we add the service parameter the service seems to work but calls from the edit service adds a second service parameter. In order to have the same format of URL for layer, search and edit it would be nice to not have to add the service parameter.

Yeah, this needs a quick fix in Admin UI so get rid of this once and for all. Will do!

**Reply by Hallbergs** [reply]
I think all issues discussed in this thread has been taken care of except for the baseLayers one. @hitomin21 @jesade-vbg were you planning to look at this or should i take care of it?

**Reply by Hallbergs** [reply]
Got an email from @jesade-vbg, they won't be able to fix this any time soon! I've got a fix coming up!

**Reply by Hallbergs** [reply]
The base-layers issue should be fixed via https://github.com/hajkmap/Hajk/commit/fb5d83f01e36a40096823a093078057e3ab2156e  (just pushed to master).

If anyone notices strange behavior regarding layers, please let me know :)

**Reply by Hallbergs** [reply]
All issues discussed here should be resolved!

**Comment by WCarlsson** [comment]
When adding a new menulink in documenthandler 2.0 in a map the name doesn´t save, the menulink is created and other settings seem to save as intended.

**Reply by Hallbergs** [reply]
This is resolved! See #920

**Comment by Hallbergs** [comment]
Just fixed another issue. See: https://github.com/hajkmap/Hajk/issues/921

**Comment by jacobwod** [comment]
Some more fixes, #920, #923 and #932. Feel free to checkout, especially @WCarlsson should like the #932 (givet that it does what I think it does…). 

_All fixes are now in `master`._

**Reply by Hallbergs** [reply]
Wonderful!!

**Reply by WCarlsson** [reply]
Awesome!

**Comment by jacobwod** [comment]
The #847 changes led to another huge bug: all of our Hajk group layers stopped working because of this: https://github.com/hajkmap/Hajk/issues/940. I've fixed it immediately but this is another example of the importance of testing.

**Comment by emmafrygne-vbg** [comment]
VARBERG 
Thank you for providing the checklist for admins. I have used it, and we are using the .Net version. We have started testing the FME plugin, but are still testing it and will do an evaluation on monday (13/12) so I let you know after that.

**Comment by maan002** [comment]
FME Server Plugin

Some small issues/features to the FME Server plugin requested by users in Varberg:

- Add previous selection to the UI, i.e. output the selected group and product in later steps.
- Be able to configure the word for "group", i.e. in Varberg they would like to use "produktkategori".
- Be able to open the product description directly after the product is selected in step 2. Maybe a button labeled "Information"?
- Be able to configure the color and fill of the selected area polygon.

**Reply by Hallbergs** [reply]
Add previous selection to the UI, i.e. output the selected group and product in later steps.
_Not sure what you mean here. Feel free to implement though._
Be able to configure the word for "group", i.e. in Varberg they would like to use "produktkategori".
_Done._
Be able to open the product description directly after the product is selected in step 2. Maybe a button labeled "Information"?
_Done._
Be able to configure the color and fill of the selected area polygon.
_Done._

And when i say done, i mean they will be in 3.8rc2 (If you want to test already, checkout #955).

**Comment by Hallbergs** [comment]
Release candidate 2 is now released! Any further discussions regarding 3.8 should take place in the discussion thread for rc2.
